pkg/datapath/tables/node_address.go

```diff
@@ -294,113 +282,147 @@ func (n *nodeAddressController) updateK8sNodeIPs(node node.LocalNode) (updated b
-	iter := n.NodeAddresses.List(txn, NodeAddressDeviceNameIndex.Query(WildcardDeviceName))
-	for addr := range iter {
-		n.NodeAddresses.Delete(txn, addr)
-	}
+	for route := range allRoutes {
+		dst := route.Dst
+		if !dst.IsValid() || dst.Addr().IsUnspecified() || dst.Addr().IsLoopback() {
+			continue
+		}
 
-	newAddrs := []NodeAddress{}
-	for _, fallback := range n.fallbackAddresses.addrs() {
-		if !fallback.IsValid() {
+		if addrsFound.Has(dst.Addr()) {
+			continue
+		}
+
+		dev, _, found := n.Devices.Get(rtxn, DeviceIDIndex.Query(route.LinkIndex))
+		if !found {
 			continue
 		}
+
+		if !n.shouldUseDeviceForNodeAddress(dev) {
+			continue
+		}
+
+		nodePort := false
+		if len(n.Config.NodePortAddresses) > 0 {
+			nodePort = dev.Name != defaults.HostDevice && ip.PrefixesContains(n.Config.NodePortAddresses, dst.Addr())
+		}
 		nodeAddr := NodeAddress{
-			Addr:       fallback,
-			NodePort:   false,
-			Primary:    true,
-			DeviceName: WildcardDeviceName,
+			Addr:       dst.Addr(),
+			NodePort:   nodePort,
+			Primary:    true, // Preferred source on a route is a strong candidate for a primary address.
+			DeviceName: dev.Name,
 		}
-		newAddrs = append(newAddrs, nodeAddr)
-		n.NodeAddresses.Insert(txn, nodeAddr)
+		newAddrsByDevice[dev.Name] = append(newAddrsByDevice[dev.Name], nodeAddr)
+		addrsFound.Insert(dst.Addr())
 	}
 
-	n.Log.Info(
-		"Fallback node addresses updated",
-		logfields.Addresses, showAddresses(newAddrs),
-		logfields.Device, WildcardDeviceName,
-	)
-}
+	// Derive wildcard addresses from fallback candidates
+	var wildcardAddrs []NodeAddress
+	for _, fallback := range n.fallbackAddresses.addrs() {
+		if fallback.IsValid() {
+			wildcardAddrs = append(wildcardAddrs, NodeAddress{
+				Addr:       fallback,
+				NodePort:   false,
+				Primary:    true,
+				DeviceName: WildcardDeviceName,
+			})
+		}
+	}
 
-func (n *nodeAddressController) updateFallbacks(txn statedb.ReadTxn, dev *Device, deleted bool) (updated bool) {
-	if dev.Name == defaults.HostDevice {
-		return false
+	if len(wildcardAddrs) > 0 {
+		newAddrsByDevice[WildcardDeviceName] = wildcardAddrs
+		n.Log.Info(
+			"Fallback node addresses updated",
+			logfields.Addresses, showAddresses(wildcardAddrs),
+			logfields.Device, WildcardDeviceName,
+		)
 	}
 
-	fallbacks := &n.fallbackAddresses
-	if deleted && fallbacks.fromDevice(dev) {
-		fallbacks.clear()
-		for dev := range n.Devices.All(txn) {
-			if strings.HasPrefix(dev.Name, "lxc") {
-				// Never pick the fallback from lxc* devices.
-				continue
-			}
-			fallbacks.update(dev)
-		}
-		return true
-	} else {
-		return n.fallbackAddresses.update(dev)
+	wtxn := n.DB.WriteTxn(n.NodeAddresses)
+	defer wtxn.Abort()
+
+	// Apply changes to the NodeAddress table
+	devicesWithAddrs := sets.New[string]()
+	for addr := range n.NodeAddresses.All(wtxn) {
+		devicesWithAddrs.Insert(addr.DeviceName)
+	}
+
+	for devName, addrs := range newAddrsByDevice {
+		n.update(wtxn, addrs, nil, devName)
```

Comment:
Should use a non-nil cell.Health here so that we see the latest update in the `health` table.